internal/convert/automation_config.go

```diff
@@ -15,61 +15,47 @@
 package convert
 
 import (
-	"fmt"
-
 	"go.mongodb.org/ops-manager/opsmngr"
 )
 
-const (
-	mongod = "mongod"
-)
+// FromAutomationConfig convert from opsmngr.AutomationConfig format to []*ClusterConfig
+func FromAutomationConfig(in *opsmngr.AutomationConfig) []*ClusterConfig {
+	out := make([]*ClusterConfig, 0, len(in.ReplicaSets))
 
-// FromAutomationConfig convert from cloud format to mCLI format
-func FromAutomationConfig(in *opsmngr.AutomationConfig) (out []ClusterConfig) {
-	out = make([]ClusterConfig, len(in.ReplicaSets))
+	for i, s := range in.Sharding {
+		out = append(out, newShardedCluster(s))
+		for j, ss := range s.Shards {
+			id := ss.ID
+			out[i].Shards[j] = newRSConfig(in, id)
+		}
 
+		out[i].Config = newRSConfig(in, s.ConfigServerReplica)
+		for j, p := range in.Processes {
+			if p.Cluster == s.Name {
+				out[i].Mongos = append(out[i].Mongos, newMongosProcessConfig(p))
+				out[i].addToMongoURI(p)
+				in.Processes = removeProcess(in.Processes, j)
+				break
+			}
+		}
+	}
 	for i, rs := range in.ReplicaSets {
-		out[i].Name = rs.ID
-		out[i].ProcessConfigs = make([]*ProcessConfig, len(rs.Members))
-
+		out = append(out, newReplicaSetCluster(rs.ID, len(rs.Members)))
 		for j, m := range rs.Members {
-			out[i].ProcessConfigs[j] = convertCloudMember(m)
 			for k, p := range in.Processes {
 				if p.Name == m.Host {
-					convertCloudProcess(out[i].ProcessConfigs[j], p)
-					if out[i].MongoURI == "" {
-						out[i].MongoURI = fmt.Sprintf("mongodb://%s:%d", p.Hostname, p.Args26.NET.Port)
-					} else {
-						out[i].MongoURI = fmt.Sprintf("%s,%s:%d", out[i].MongoURI, p.Hostname, p.Args26.NET.Port)
-					}
-					in.Processes = append(in.Processes[:k], in.Processes[k+1:]...)
+					out[i].ProcessConfigs[j] = newReplicaSetProcessConfig(m, p)
+					out[i].addToMongoURI(p)
+					in.Processes = removeProcess(in.Processes, k)
```

Comment:
since this is such a complex search (`O(n*m*o)`) I'm trying to speed up search by deleting visited processes 
I want at some point benchmark this to see how else can we improve it

Reply:
ah got it

Reply:
I changed the name from in and added a comment about how this method is dangerous